Renesas RX Zephyr architecture support

duynguyenxa opened the pull request:
This is the initial PR to support for RX architecture.
Renesas RX core is a proprietary architecture from Renesas.
This PR include the support of:
- RXv1 core architecture basic support
- SOC: [RX130 MCU](https://www.renesas.com/en/products/microcontrollers-microprocessors/rx-32-bit-performance-efficiency-mcus/rx130-cost-optimized-high-performance-32-bit-microcontroller-enhanced-touch-key-function-and-5v-operation?srsltid=AfmBOopf6QvGGPa3K0kr6ugZ_yfphdd68XnEJW6rOyUb9gpAYCRgjduJ)
- Boards: [RSKRX130 512KB Board](https://www.renesas.com/en/products/microcontrollers-microprocessors/rx-32-bit-performance-efficiency-mcus/rx130-512kb-starter-kit-renesas-starter-kit-rx130-512kb)
- RX qemu support
- Driver:
  - Clock control (CGC)
  - System timer (CMT)
  - GPIO
  - Pinctrl
  - Serial
  
How to run test on this architecture qemu environment:
- Download Renesas RX GCC toolchain: https://llvm-gcc-renesas.com/rx-download-toolchains/
- Configure Zephyr build environment variable to use gcc for RX:
`export ZEPHYR_TOOLCHAIN_VARIANT=cross-compile`
`export RXGCC_TOOLCHAIN_PATH=<PATH-to-GCC-RX-toolchain/bin/rx-elf->`
- Build and run sample on qemu_rx target:
`west build -b qemu_rx -p always sample/synchronization`
`west build -t run`


Related ticket: https://github.com/zephyrproject-rtos/zephyr/issues/81313

duynguyenxa committed:
goc: renesas: rx: Initial support for RX130 SOC

Minimal SOC layer support for Renesas RX SOC
This SOC is using Renesas RXv1 core

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
boards: renesas: Add support for RSKRX130 512KB boards

Intial support for RSKRX130 512KB board
The board 2 version 128KB and 512KB running on 2 different
part of the RX130 MCU series.

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

Commit:
scripts: west: flash: Add support for .mot file format

Add support for .mot file flash using west flash command
The RX build output .mot as binary file to flash into
board

Signed-off-by: Phi Tran <phi.tran.jg@bp.renesas.com>
Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
drivers: clock: Support clock control driver RX MCU

Initial support of clock control driver for RX MCU

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
Signed-off-by: Tran Van Quy <quy.tran.pz@renesas.com>

duynguyenxa committed:
drivers: serial: Initial support for Renesas RX serial driver

Intial serial driver support for RX MCU, this driver utilize
the SCI HWIP for uart communication
Current support include polling API and Interrupt driven API,
some of the code is using Renesas RX Driver Package (RDP) as
hal layer

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
Signed-off-by: Sang Tran <sang.tran.jc@renesas.com>

duynguyenxa committed:
drivers: pinctrl: Support pinctrl driver for Renesas RX

Intial support of pinctrl driver for Renesas RX MCU
family.
This support base on using Renesas RX driver package in
hal_renesas layer

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
Signed-off-by: Phi Tran <phi.tran.jg@bp.renesas.com>

duynguyenxa committed:
drivers: gpio: Support GPIO driver for Renesas RX MCU

Initial commit for GPIO driver support on board using RX130 MCUs
* drivers: GPIO: implementation for GPIO driver on RSK_RX130_512KB
* dts: rx: add device node for GPIO of RSK_RX130_512KB

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
Signed-off-by: Phi Tran <phi.tran.jg@bp.renesas.com>

nordicjm commented on cmake/compiler/gcc/target_rx.cmake:
```diff
@@ -0,0 +1,25 @@
+  list(APPEND TOOLCHAIN_LD_FLAGS)
+  list(APPEND TOOLCHAIN_LD_FLAGS -mlittle-endian-data)
+
+  if (NOT CONFIG_FPU)
+    list(APPEND TOOLCHAIN_C_FLAGS -nofpu)
+  endif()
+
+  if (CONFIG_CPU_RXV1)
+    list(APPEND TOOLCHAIN_C_FLAGS -misa=v1)
+    list(APPEND TOOLCHAIN_LD_FLAGS -misa=v1)
+  elseif (CONFIG_CPU_RXV2)
+    list(APPEND TOOLCHAIN_C_FLAGS -misa=v2)
+    list(APPEND TOOLCHAIN_LD_FLAGS -misa=v2)
+  else ()
```
```suggestion
  if(NOT CONFIG_FPU)
    list(APPEND TOOLCHAIN_C_FLAGS -nofpu)
  endif()

  if(CONFIG_CPU_RXV1)
    list(APPEND TOOLCHAIN_C_FLAGS -misa=v1)
    list(APPEND TOOLCHAIN_LD_FLAGS -misa=v1)
  elseif(CONFIG_CPU_RXV2)
    list(APPEND TOOLCHAIN_C_FLAGS -misa=v2)
    list(APPEND TOOLCHAIN_LD_FLAGS -misa=v2)
  else()
```

duynguyenxa replied:
It's fixed

nordicjm commented on cmake/toolchain/rxgcc/generic.cmake:
```diff
@@ -0,0 +1,38 @@
+ endif()
+
+set_ifndef(RXGCC_TOOLCHAIN_PATH "$ENV{RXGCC_TOOLCHAIN_PATH}")
+set(       RXGCC_TOOLCHAIN_PATH ${RXGCC_TOOLCHAIN_PATH} CACHE PATH "rx gcc install directory")
+assert(    RXGCC_TOOLCHAIN_PATH "RXGCC_TOOLCHAIN_PATH is not set")
```
```suggestion
set(RXGCC_TOOLCHAIN_PATH ${RXGCC_TOOLCHAIN_PATH} CACHE PATH "rx gcc install directory")
assert(RXGCC_TOOLCHAIN_PATH "RXGCC_TOOLCHAIN_PATH is not set")
```

duynguyenxa replied:
It's fixed

nordicjm commented on cmake/toolchain/rxgcc/generic.cmake:
```diff
@@ -0,0 +1,38 @@
+# Search for rxgcc toolchain
+if(("rxgcc" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) AND (DEFINED ZEPHYR_TOOLCHAIN_VARIANT))
+  # No toolchain was specified, so inform user that we will be searching.
+  if (NOT DEFINED ENV{RXGCC_TOOLCHAIN_PATH})
```
```suggestion
  if(NOT DEFINED ENV{RXGCC_TOOLCHAIN_PATH})
```

duynguyenxa replied:
It's fixed

nordicjm commented on include/zephyr/arch/rx/linker.ld:
```diff
@@ -0,0 +1,247 @@
+#include <zephyr/linker/linker-tool.h>
+
+#ifdef CONFIG_XIP
+#define ROMABLE_REGION 				ROM
```
doesn't align with below

duynguyenxa replied:
It's fixed

nordicjm commented on include/zephyr/arch/rx/linker.ld:
```diff
@@ -0,0 +1,247 @@
+	OFS (r)  : ORIGIN = 0xFE7F5D00, LENGTH = 256
+
+	LINKER_DT_REGIONS()
+    /* Used by and documented in include/linker/intlist.ld */
+    IDT_LIST (wx) : ORIGIN = 0xFFFFF780, LENGTH = 2K
```
not using tab

duynguyenxa replied:
It's fixed

nordicjm commented on include/zephyr/arch/rx/linker.ld:
```diff
@@ -0,0 +1,247 @@
+ENTRY(CONFIG_KERNEL_ENTRY)
+
+SECTIONS
+	{
```
indent

duynguyenxa replied:
It's fixed

nordicjm commented on include/zephyr/toolchain/gcc.h:
```diff
@@ -575,6 +575,17 @@ do {                                                                    \
+		"\n\t.type\t" #name ",%%object" :  : "n"(value))
+
+#define GEN_ABSOLUTE_SYM_KCONFIG(name, value)       \
+	__asm__(".globl\t" #name                    \
```
global typo?

duynguyenxa replied:
Thank you I change it to ".global"

nordicjm commented on dts/bindings/flash_controller/renesas,rx-flash.yaml:
```diff
@@ -0,0 +1,53 @@
+
+compatible: "renesas,rx-flash"
+
+include: base.yaml
```
`include: flash-controller.yaml` ?

duynguyenxa replied:
I add new binding in mtd for flash node and remove redundant properties

nordicjm commented on dts/bindings/flash_controller/renesas,rx-flash.yaml:
```diff
@@ -0,0 +1,53 @@
+
+include: base.yaml
+
+properties:
```
half of these aren't needed if the right file is included

duynguyenxa replied:
I add new binding in mtd and remove redundant properties

nordicjm commented on dts/rx/renesas/r5f513083xfb.dtsi:
```diff
@@ -0,0 +1,124 @@
+				reg = <0xfff80000 DT_SIZE_K(512)>;
+				reg-names = "OFFSET";
+				entry_bit = <0>;
+				write_block_size = <128>;
```
add erase block size

duynguyenxa replied:
Modified write block size and added erase block size according to HWM

nordicjm commented on dts/rx/renesas/r5f513083xfb.dtsi:
```diff
@@ -0,0 +1,124 @@
+				erased_undefined;
+				reg = <0x00100000 DT_SIZE_K(8)>;
+				reg-names = "OFFSET";
+				entry_bit = <7>;
```
add erase block size

duynguyenxa replied:
Same as above

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -0,0 +1,87 @@
+			reg = <0>;
+			status = "okay";
+		};
+	}; /* cpus */
```
```suggestion
	};
```

duynguyenxa replied:
It's removed

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -0,0 +1,87 @@
+		      <0x0087300 0xff>,
+		      <0x00872f0 0x02>;
+		reg-names = "IR", "IER", "IPR", "FIR";
+	}; /* icu */
```
as above

duynguyenxa replied:
It's removed

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -0,0 +1,87 @@
+			zephyr,memory-region = "OFSM";
+			status = "okay";
+		};
+	}; /* soc */
```
as above

duynguyenxa replied:
It's removed

nordicjm commented on include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-rx.h:
```diff
@@ -0,0 +1,10 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_PINCTRL_RX_H_
```
why is this empty file needed?

duynguyenxa replied:
This file is removed in the commit thank you

nordicjm commented on soc/renesas/rx/rx130/Kconfig.defconfig:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_SERIES_RX130
+
```
file can go

duynguyenxa replied:
This file is removed in the commit

nordicjm commented on soc/renesas/rx/rx130/Kconfig.defconfig:
```diff
@@ -3,4 +3,7 @@
 
 if SOC_SERIES_RX130
 
+config PINCTRL
+	default y
```
no

duynguyenxa replied:
The config is removed and select in driver config

nordicjm commented on boards/renesas/rsk_rx130_512kb/rsk_rx130_512kb_defconfig:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+# enable GPIO
+CONFIG_GPIO=y
+
+# enable uart driver
+CONFIG_SERIAL=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+
+# enable console
```
```suggestion
# Enable GPIO
CONFIG_GPIO=y

# Enable uart driver
CONFIG_SERIAL=y
CONFIG_UART_INTERRUPT_DRIVEN=y

# Enable console
```

duynguyenxa replied:
It's fixed

nordicjm commented on boards/renesas/rsk_rx130_512kb/rsk_rx130_512kb_defconfig:
```diff
@@ -0,0 +1,14 @@
+# enable console
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+CONFIG_DEBUG=y
```
remove

duynguyenxa replied:
The CONFIG_DEBUG is remove thank you

nordicjm commented on drivers/timer/renesas_rx_cmt.c:
```diff
@@ -0,0 +1,136 @@
+#define MAX_TICKS       ((COUNTER_MAX / CYCLES_PER_TICK) - 1)
+
+#if (IS_ENABLED(CONFIG_TICKLESS_KERNEL))
+#warning Warning: Tickless mode is not yet implemented.
```
if it's not implemented then why is this a warning and not an error?

duynguyenxa replied:
It's changed to #error

nordicjm commented on drivers/timer/renesas_rx_cmt.c:
```diff
@@ -0,0 +1,136 @@
+	sys_clock_announce(dticks);
+}
+
+#define CMT0_IRQ_NUM DT_IRQ_BY_NAME(CMT0_NODE, cmi, irq)
```
move defines to top of file with other defines

duynguyenxa replied:
It's fixed

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -395,13 +395,37 @@
+				compatible = "renesas,rx-timer-cmt";
+				#address-cells = <1>;
+				#size-cells = <1>;
+				reg =	<0x00088002 0x02>,
+					<0x00088004 0x02>,
+					<0x00088006 0x02>;
```
remove double space, use space after tables for alignment on other lines if needed

duynguyenxa replied:
It's fixed

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -395,13 +395,37 @@
+			compatible = "renesas,rx-timer-cmt-start-control";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			reg =	<0x00088000 0x02>;
```
single space

duynguyenxa replied:
It's fixed

nordicjm replied:
not fixed

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -395,13 +395,37 @@
+
+			cmt1: timer@88008 {
+				compatible = "renesas,rx-timer-cmt";
+				reg =	<0x00088008 0x02>,
+					<0x0008800A 0x02>,
+					<0x0008800C 0x02>;
```
as above

duynguyenxa replied:
It's fixed

nordicjm commented on boards/renesas/rsk_rx130_512kb/Kconfig.rsk_rx130_512kb:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_RSK_RX130_512KB
```
why have you used 512kb in the name, is this the actual name of the board you sell "rsk rx130 512kb"?

duynguyenxa replied:
@nordicjm , yes on the web site its name is Renesas Starter Kit RX130 512kb.
https://www.renesas.com/en/products/microcontrollers-microprocessors/rx-32-bit-performance-efficiency-mcus/rx130-512kb-starter-kit-renesas-starter-kit-rx130-512kb?doc_secondary=visible

There is another board name RSK RX130 which only has 128kb of ROM
https://www.renesas.com/en/products/microcontrollers-microprocessors/rx-32-bit-performance-efficiency-mcus/rx130-128kb-starter-kit-renesas-starter-kit-rx130-128kb-non-promotion?srsltid=AfmBOorl8HwnTB3GbXUaUvPV0ZDL4lwUP2lzebUCAQKsLAsye_hO6Qo4

We also use this naming to be unify with the Renesas e2studio naming of the board
![image](https://github.com/user-attachments/assets/cb3829df-be91-49cd-bcd4-975c10c1eec9)

nordicjm replied:
There are board revisions in zephyr, so you could have `rsk_rx130@512kb` and `rsk_rx130@128kb` if you wanted to which means you can have common files e.g. devicetree then have overlayes which provide the different parts, or they can be kept as separate boards.

duynguyenxa replied:
hi @nordicjm , I've update the board to rsk_rx130@512kb

nashif commented:
@duynguyenxa  is there Qemu support for this architecture? how will this be validated in CI?

nashif commented:
> @duynguyenxa is there Qemu support for this architecture? how will this be validated in CI?

i see this https://www.qemu.org/docs/master/system/target-rx.html, will you add this in the initial port?

carlescufi commented:
Hi @duynguyenxa. Would you be willing and able to present this in the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group)? If so, please let me know here and I will schedule a date for you.

duynguyenxa committed:
boards: qemu_rx: Add support for qemu rx

This support is based on the RX62N qemu envinronment
See: https://www.qemu.org/docs/master/system/target-rx.html

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
> > @duynguyenxa is there Qemu support for this architecture? how will this be validated in CI?
> 
> i see this https://www.qemu.org/docs/master/system/target-rx.html, will you add this in the initial port?

Yes, I will add the qemu target in this initial port. Currently some Kernel tests are still fail on the qemu environment.

duynguyenxa commented:
> Hi @duynguyenxa. Would you be willing and able to present this in the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group)? If so, please let me know of your availability here and I will schedule a date for you.

@carlescufi , What should be the content of the present? Is it reviewing this code? Can you help to list the content or material that I need to prepare for the present. Thank you.

duynguyenxa committed:
soc: renesas: Add support for RX62N MCU

The qemu-system-rx is based on RX62N, this commit added
support for the RX62N SOC layer. MCU is using RXv1 core and
system timer running at 6MHz

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
twister: Add support for new rx arch test

Add "rx" in the arch list of twister

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
test: Test accumulator register save for RX arch

Add new test cases to check for accumulator register save
on RX architecture

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
driver: serial: Add support for uart interface for qemu_rx

Support uart driver for qemu_rx environment base on the SCI0
HW on RX MCU

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
driver: timer: Support for RX system timer

This commit add a system timer driver for Renesas RX using the
CMT peripheral. The driver supports both system ticks and
high-resolution cycle counting
- Configures CMT0 as the system tick timer
- Configures CMT1 as a free-running cycle timer for precise
  time tracking
- Handles timer overflows to maintain a continuous cycle count.
- Implements sys_clock_cycle_get_32() and sys_clock_cycle_get_64()
  for  high-resolution timing
- Supports Zephyr tickless kernel mode by tracking elapsed cycles
- Enables interrupt-based tick announcement using CMT0

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>
Signed-off-by: Yuichi Nakada <yuichi.nakada.sx@renesas.com>

duynguyenxa committed:
test: fatal_exception: Add support for RX test

Add RX exception instruction

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
test: interrupt_utils: Add method to trigger interrupt on RX

Add trigger_irq() function support for RX
This function will set corresponding IR bit to 1 on RX ICU to
trigger interrupt

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
Last push includes:
+ Update for system timer driver for tickless operation by using 2 CMT channel, 64-bit timer get.
+ Change the implementation of the irq_offload.
+ Add qemu envinronment for the RX CPU based on the RX62N

carlescufi commented:
> > Hi @duynguyenxa. Would you be willing and able to present this in the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group)? If so, please let me know of your availability here and I will schedule a date for you.
> 
> @carlescufi , What should be the content of the present? Is it reviewing this code? Can you help to list the content or material that I need to prepare for the present. Thank you.

Hi @duynguyenxa. This would be really up to you. You can prepare slides or just talk over the code directly, whatever you think works best. The idea is just to present this work to more contributors so they can get a better idea bout the work.

I can propose either March 11th (very short notice sorry), March 18th or March 25th

duynguyenxa commented:
> > > Hi @duynguyenxa. Would you be willing and able to present this in the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group)? If so, please let me know of your availability here and I will schedule a date for you.
> > 
> > 
> > @carlescufi , What should be the content of the present? Is it reviewing this code? Can you help to list the content or material that I need to prepare for the present. Thank you.
> 
> Hi @duynguyenxa. This would be really up to you. You can prepare slides or just talk over the code directly, whatever you think works best. The idea is just to present this work to more contributors so they can get a better idea bout the work.
> 
> I can propose either ~March 11th (very short notice sorry),~ March 18th or March 25th

@carlescufi , I can do the present on March 25th, can you help to schedule me on that day. Thank you

carlescufi commented:
> > > > Hi @duynguyenxa. Would you be willing and able to present this in the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group)? If so, please let me know of your availability here and I will schedule a date for you.
> > > 
> > > 
> > > @carlescufi , What should be the content of the present? Is it reviewing this code? Can you help to list the content or material that I need to prepare for the present. Thank you.
> > 
> > 
> > Hi @duynguyenxa. This would be really up to you. You can prepare slides or just talk over the code directly, whatever you think works best. The idea is just to present this work to more contributors so they can get a better idea bout the work.
> > I can propose either ~March 11th (very short notice sorry),~ March 18th or March 25th
> 
> @carlescufi , I can do the present on March 25th, can you help to schedule me on that day. Thank you

@duynguyenxa great! I confirm that you are scheduled for March 25th.

duynguyenxa committed:
CMake: Add support for .mot binary file format

The Renesas RX support flashing .mot file for binary image
This commit target to add the .mot file output for build and
flash script

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
cmake: Add Renesas RX support

This commit add toolchain command line option for the RX
architecture support

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
include: Update toolchain header to support RX architecture

This add common gcc define for new RX architecture support

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
include: Create linker for RX architecture

Add intial support for RX linker script
The RX GCC toolchain automatically adding an "_" as prefix
on every C symbol to comply to the arch ABI, for the C file
to understand the symbol define in linker script the
PLACE_SYMBOL_HERE() macro help to add "_" to linker script
define symbol if build with CONFIG_RX enable
This commit adding minimal support in common linker script
for RX arch

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa committed:
arch: Initial support for RX architecture support

This commit add require code for supporting RX architecture
to Zephyr, it include:
- Add require config and CMakelists for RX arch
- Intialization startup code for RX
- Interrupt and exception handling
- Thread creation adn thread context switch
- irq offload using SW interrupt

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

henrikbrixandersen commented:
Architecture WG meeting 2025-03-25:
- @duynguyenxa gave a thorough presentation of the proposed Renesas RX architecture port (are the slides available online?)
- The need for a custom GCC-based toolchain was discussed. The idea is to submit the needed patches to the Zephyr SDK GCC.

duynguyenxa committed:
test: Skip kernel stack protect test for RX arch

Skip kernel stack protect as the compiler not support
-fstack-protector

Signed-off-by: Duy Nguyen <duy.nguyen.xa@renesas.com>

duynguyenxa commented:
hi @nashif , I've updated the branch to use with "cross-compile" to build the RX architecture.
Current kernel test result: 
On qemu_rx: **11 failed cases**

> INFO    - Total complete:  130/ 130  100%  built (not run):    2, filtered:  106, failed:   10, error:    1

INFO    - 200 test scenarios (201 configurations) selected, 106 configurations filtered (71 by static filter, 35 at runtime).
INFO    - 82 of 95 executed test configurations passed (86.32%), 2 built (not run), 10 failed, 1 errored, with no warnings in 695.66 seconds.
INFO    - 1015 of 1035 executed test cases passed (98.07%), 10 blocked, 10 failed on 1 out of total 1007 platforms (0.10%).
INFO    - 132 selected test cases not executed: 130 skipped, 2 not run (built only).
INFO    - 92 test configurations executed on platforms, 3 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/duyhoangnguyen/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) tests/kernel/fifo/fifo_api/kernel.fifo on qemu_rx/r5f562n8 failed (Timeout)
INFO    - 2) tests/kernel/workq/critical/kernel.workqueue.critical on qemu_rx/r5f562n8 failed (Testsuite failed)
INFO    - 3) tests/kernel/interrupt/arch.shared_interrupt on qemu_rx/r5f562n8 failed (Timeout)
INFO    - 4) tests/kernel/timer/timer_api/kernel.timer on qemu_rx/r5f562n8 failed (Testsuite failed)
INFO    - 5) tests/kernel/tickless/tickless_concept/kernel.tickless.concept on qemu_rx/r5f562n8 failed (Testsuite failed)
INFO    - 6) tests/kernel/sched/schedule_api/kernel.scheduler on qemu_rx/r5f562n8 failed (Timeout)
INFO    - 7) tests/kernel/sched/schedule_api/kernel.scheduler.slice_perthread on qemu_rx/r5f562n8 failed (Testsuite failed)
INFO    - 8) tests/kernel/sched/schedule_api/kernel.scheduler.simple_timeslicing on qemu_rx/r5f562n8 failed (Timeout)
INFO    - 9) tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random on qemu_rx/r5f562n8 error (Build failure - error: unknown option after #pragma GCC diagnostic kind [-Werror=pragmas])
INFO    - 10) tests/kernel/semaphore/semaphore/kernel.semaphore on qemu_rx/r5f562n8 failed (Timeout)

On actual HW RSK-RX130 board: **All pass**

> INFO    - Using 'cross-compile' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/builds/zephyr-bsp-dev/zephyr/twister-out/testplan.json
INFO    - JOBS: 48
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:  132/ 132  100%  built (not run):   97, filtered:  104, failed:    0, error:    0
INFO    - 200 test scenarios (201 configurations) selected, 104 configurations filtered (69 by static filter, 35 at runtime).
INFO    - 0 of 97 executed test configurations passed (0.00%), 97 built (not run), 0 failed, 0 errored, with no warnings in 158.99 seconds.
INFO    - 0 test configurations executed on platforms, 97 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/duyhoangnguyen/builds/zephyr-bsp-dev/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/duyhoangnguyen/builds/zephyr-bsp-dev/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/duyhoangnguyen/builds/zephyr-bsp-dev/zephyr/twister-out/twister_report.xml...
INFO    - Run completed

> DEBUG   - Release DUT:rsk_rx130@512kb/r5f513083xfb, Id:None, counter:95, failures:0
DEBUG   - run status: rsk_rx130@512kb/r5f513083xfb/cross-compile/tests/kernel/common/kernel.common.minimallibc passed
INFO    - Total complete:   95/  95  100%  built (not run):    0, filtered:    0, failed:    0, error:    0
INFO    - 3587 test scenarios (95 configurations) selected, 0 configurations filtered (0 by static filter, 0 at runtime).
INFO    - 95 of 95 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 1691.28 seconds.
INFO    - 1041 of 1154 executed test cases passed (90.21%), 113 blocked on 1 out of total 985 platforms (0.10%).
INFO    - 134 selected test cases not executed: 134 skipped.
INFO    - 95 test configurations executed on platforms, 0 test configurations were only built.
Hardware distribution summary:
| Board                        | ID   |   Counter |   Failures |
|------------------------------|------|-----------|------------|
| rsk_rx130@512kb/r5f513083xfb |      |        95 |          0 |

nashif commented:
@nordicjm  Please revisit.

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -0,0 +1,87 @@
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		cpu0: cpu@0 {
```
newline

duynguyenxa replied:
newline is added

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -0,0 +1,87 @@
+			compatible = "renesas,rx-sci";
+			reg = <0x0008A005 0x100>;
+			status = "disabled";
+			uart {
```
newline

duynguyenxa replied:
newline is added

nordicjm commented on soc/renesas/rx/rx130/Kconfig.defconfig:
```diff
@@ -0,0 +1,20 @@
+# Set the PCLKB/8 frequency to SYS_CLOCK_HW_CYCLES_PER_SEC.
+# Because the CMT of Systick Timer divides the source clock of PCLKB by 8-divider.
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 4000000
```
this should come from a dts property

duynguyenxa replied:
I added new clock-freq property in the sysclock CMT node and change the config to use that dts property

nordicjm commented on dts/rx/renesas/rx-qemu.dtsi:
```diff
@@ -0,0 +1,110 @@
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+	cpus {
```
newline

nordicjm commented on dts/rx/renesas/rx-qemu.dtsi:
```diff
@@ -0,0 +1,110 @@
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		cpu0: cpu@0 {
```
newline etc. apply to whole PR, not marking rest

duynguyenxa replied:
this dts file is formatted. Thank you

nordicjm commented on dts/rx/renesas/rx-qemu.dtsi:
```diff
@@ -0,0 +1,110 @@
+		      <0x00872f0 0x02>;
+		reg-names = "IR", "IER", "IPR", "FIR";
+	};
+
```
```suggestion
```

duynguyenxa replied:
Newline is removed

nordicjm commented on soc/renesas/rx/rx62n/Kconfig.defconfig:
```diff
@@ -0,0 +1,20 @@
+# Set the PCLKB/8 frequency to SYS_CLOCK_HW_CYCLES_PER_SEC.
+# Because the CMT of Systick Timer divides the source clock of PCLKB by 8-divider.
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 6000000
```
as above

duynguyenxa replied:
I added new clock-freq property in the sysclock CMT node and change the config to use that dts property

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -58,9 +59,19 @@
 		};
 
-		sci0: sci0@8a005 {
+		sci0: sci0@8a000 {
```
seems the original commit is wrong?

duynguyenxa replied:
Yes, thank you for catching this, the address of the node is incorrect in the original commit, I've fixed it.

nordicjm commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -387,8 +391,68 @@
+			clocks = <&pclkb MSTPB 30>;
+			status = "disabled";
+			channel = <1>;
+			uart {
```
here too

duynguyenxa replied:
newline is added

nordicjm commented on arch/rx/Kconfig:
```diff
@@ -91,12 +91,6 @@ config MAX_IRQ_LINES # check if: TODO: adaptation required
-	default 120000000
-	help
-	  This option specifies hardware clock.
-
```
not really sure why this was added in a commit of this PR, and is now being removed in the same PR in a later commit?

duynguyenxa replied:
This is due to bad rebase, I've fixed it, thank you

nordicjm commented on boards/renesas/rsk_rx130/rsk_rx130.dts:
```diff
@@ -0,0 +1,76 @@
+	};
+
+	leds {
+		compatible = "gpio-leds";
+		led1: led1 {
```
newlines needed in this commit too

thaoluonguw replied:
Add blank line

nordicjm commented on boards/qemu/rx/qemu_rx.dts:
```diff
@@ -0,0 +1,51 @@
+
+&sci0 {
+	status = "okay";
+	uart0: uart {
```
as above

nordicjm commented on boards/qemu/rx/qemu_rx_defconfig:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+# enable uart driver
```
capital first letters and uppercase acronyms e.g. UART

thaoluonguw commented on boards/renesas/rsk_rx130/doc/index.rst:
```diff
@@ -0,0 +1,155 @@
+About ``Install dependencies`` and ``Get Zephyr and install Python dependencies``,
+please visit `Zephyr Getting Started`_ to install dependencies.
+
+For RX MCU build on Zephyr, currently Zephyr SDK is not supported we need to install
```
"is not supported" -> "haven't supported"

thaoluonguw commented on boards/renesas/rsk_rx130/doc/index.rst:
```diff
@@ -0,0 +1,155 @@
+.. _Renesas Debug extension:
+   https://marketplace.visualstudio.com/items?itemName=RenesasElectronicsCorporation.renesas-debug
+
+.. _Zephyr Getting Started:
+   https://docs.zephyrproject.org/latest/develop/getting_started/index.html
```
Do we need for this reference link?

duynguyenxa replied:
I removed it

thaoluonguw commented on boards/renesas/rsk_rx130/rsk_rx130.dts:
```diff
@@ -0,0 +1,76 @@
+&sci1 {
+	pinctrl-0 = <&sci1_default>;
+	pinctrl-names = "default";
+	status = "okay";
+	uart1: uart {
```
Add blank line

duynguyenxa commented on cmake/compiler/gcc/target_rx.cmake:
```diff
@@ -0,0 +1,38 @@
+  list(APPEND TOOLCHAIN_LD_FLAGS)
+  list(APPEND TOOLCHAIN_LD_FLAGS -mlittle-endian-data)
+
+  if (NOT CONFIG_PICOLIBC)
```
I've removed the space

duynguyenxa commented on cmake/compiler/gcc/target_rx.cmake:
```diff
@@ -0,0 +1,38 @@
+    list(APPEND TOOLCHAIN_C_FLAGS -nofpu)
+  endif()
+
+  if ("cross-compile" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT})
```
I've removed the space

duynguyenxa commented on cmake/compiler/gcc/target_rx.cmake:
```diff
@@ -0,0 +1,38 @@
+      list(APPEND TOOLCHAIN_LD_FLAGS -misa=v3)
+    endif()
+  elseif("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT})
+    if (CONFIG_SOC_SERIES_RX130)
```
I've removed the space

duynguyenxa commented on arch/rx/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+
+add_subdirectory(core)
+  set_property(GLOBAL PROPERTY PROPERTY_OUTPUT_FORMAT elf32-rx-le) # needed for e.g. objcopy
+  set_property(GLOBAL PROPERTY PROPERTY_OUTPUT_ARCH "v2")
```
I've removed the indent

duynguyenxa commented on dts/rx/renesas/rx130-common.dtsi:
```diff
@@ -0,0 +1,87 @@
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+	cpus {
```
newline is added

duynguyenxa commented on arch/rx/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+
+add_subdirectory(core)
+  set_property(GLOBAL PROPERTY PROPERTY_OUTPUT_FORMAT elf32-rx-le) # needed for e.g. objcopy
+  set_property(GLOBAL PROPERTY PROPERTY_OUTPUT_ARCH "v2")
```
Hi @thaoluonguw , thank you for checking, the original port is for RX65N which is using the RXv2 core, this property is not use anywhere so I removed it.

duynguyenxa commented on arch/rx/Kconfig:
```diff
@@ -0,0 +1,127 @@
+	  and a slightly increased boot time.
+
+	  If the CPU's d-cache line size is known in advance, disable this
+	  option and manually enter the value for CACHE_LINE_SIZE.
```
This port current does not support cache on RX core so I removed the config, thank you

duynguyenxa commented on arch/rx/Kconfig:
```diff
@@ -0,0 +1,127 @@
+	default 32
+	help
+	  Size in bytes of a CPU d-cache line. Detect automatically at runtime
+	  by selecting CACHE_LINE_SIZE_DETECT.
```
yes, but on this PR we does not support cache on RX core so I removed the config

duynguyenxa commented on arch/rx/Kconfig:
```diff
@@ -0,0 +1,127 @@
+	range 1 16
+	default 16
+	help
+	  Interrupt priorities available will be 0 to NUM_IRQ_PRIO_LEVELS-1. The
+	  minimum value is 1. The BSP must provide a valid default for proper
+	  operation.
```
The RX ICU (Interrupt control unit) priority is from 0:15 (0 is lowest, 15 is highest), this config this to define the total number of irq (0 to 15 is 16).

duynguyenxa commented on arch/rx/Kconfig:
```diff
@@ -0,0 +1,127 @@
+config GEN_SW_ISR_TABLE
+	default y
+
+config GEN_IRQ_VECTOR_TABLE
+	default n
```
No it's not, the HAS_EXCEPT_VECTOR_TABLE is related to the "extb" handling in RX core HW.
This config is to prevent the Zephyr build system to generate the HW interrupt table, this is due to the limitation of RX core does not have an appropriate way to demux the interrupt on the core, so we have to implement the table in the vect.c file

duynguyenxa commented on arch/rx/Kconfig:
```diff
@@ -0,0 +1,127 @@
+config MAX_IRQ_LINES # check if: TODO: adaptation required
+	int "Number of IRQ lines"
+	default 128
+	range 0 224
```
This config is a reference from x86 architecture and is a to do work to improve the interrupt table generation to help reduce the size of the code, now it has no impact on the build so I remove this.

duynguyenxa commented on arch/rx/Kconfig:
```diff
@@ -0,0 +1,127 @@
+	  Unused Pins" section of PORT chapter of RX MCU of User's manual.
+	  Note: please MUST set "BSP_PACKAGE_PINS" definition to your device
+	  of pin type in r_bsp_config.h.
+	  Otherwise, the port may output without intention.
```
This config is to prevent the un-expect output from port, it's depended on HAL layer and the code implementation to the "arch" so I moved this to SOC Kconfig and the soc early init hook

nordicjm reviewed (commented):
Some small nit. board/soc changes OK

duynguyenxa commented on CMakeLists.txt:
```diff
@@ -1837,6 +1837,27 @@ if(CONFIG_BUILD_OUTPUT_BIN AND CONFIG_BUILD_OUTPUT_UF2)
+      post_build_byproducts
+      ${KERNEL_MOT_NAME}
+      )
+  endif()
```
I added it

duynguyenxa commented on soc/renesas/rx/Kconfig:
```diff
@@ -0,0 +1,21 @@
+
+if SOC_FAMILY_RENESAS_RX
+
+config NONE_USED_PORT_INIT
```
I change the config to RENESAS_NONE_USED_PORT_INIT

duynguyenxa commented on boards/renesas/rsk_rx130/rsk_rx130_defconfig:
```diff
@@ -0,0 +1,13 @@
+# Enable GPIO
+CONFIG_GPIO=y
+
+# Enable uart driver
```
The comment is update to UART

nordicjm commented on drivers/pinctrl/renesas/rx/pinctrl_renesas_rx.c:
```diff
@@ -0,0 +1,153 @@
+
+		port_pin = (pin->port_num << PORT_POS) | pin->pin_num;
+
+		/* set PMR register to 0 before setting pin control register*/
```
missing space at end of all these comments, should all start with capital letter too

duynguyenxa replied:
Thank you for checking, I've fixed all the comment in this file.

nordicjm commented on boards/renesas/rsk_rx130/doc/index.rst:
```diff
@@ -0,0 +1,147 @@
+                "deviceFamily": "RX",
+                "device": "R5F51308",
+                "debuggerType": "SEGGERJLINKRX",
+        }
```
missing } for end of target block

duynguyenxa replied:
Thank you, I've added the missing }

kartben commented on boards/renesas/rsk_rx130/board.yml:
```diff
@@ -0,0 +1,12 @@
+board:
+  name: rsk_rx130
```
Please add `full_name`

duynguyenxa replied:
Thank you for checking, I added `full_name: Renesas Starter Kit for RX130`

kartben commented on boards/qemu/rx/board.yml:
Board is missing a documentation page

duynguyenxa replied:
I've added description for qemu_rx environment

kartben commented on boards/renesas/rsk_rx130/doc/index.rst:
```diff
@@ -0,0 +1,147 @@
+Supported Features
+==================
+
+The below features are currently supported on Zephyr for RSK-RX130-512KB board:
+
+.. zephyr:board-supported-hw::
+
+Other hardware features are currently not supported by the port.
```
```suggestion
.. zephyr:board-supported-hw::
```

duynguyenxa replied:
This is fixed as suggestion

kartben commented on boards/renesas/rsk_rx130/doc/index.rst:
```diff
@@ -0,0 +1,147 @@
+Applications for the ``rsk_rx130@512kb`` board target configuration can be
+built, flashed, and debugged as below.
+
+Currently, the Zephyr SDK hasn't added support for RX builds yet, so the GCC for RX toolchain is required and build system need to be set to use "cross-compile".
```
From this point until the end of the file, indenting needs some love as rendering is quite broken :)

https://builds.zephyrproject.io/zephyr/pr/81507/docs/boards/renesas/rsk_rx130/doc/index.html#programming-and-debugging

duynguyenxa replied:
Thank you, I've reformatted the document

kartben commented on boards/renesas/rsk_rx130/doc/index.rst:
```diff
@@ -0,0 +1,147 @@
+Hardware
+********
+Detailed hardware features can be found at:
+- RX130 MCU: `RX130 Group User's Manual Hardware`_
```
```suggestion

- RX130 MCU: `RX130 Group User's Manual Hardware`_
```

duynguyenxa replied:
This is fixed as suggestion

kartben reviewed (changes requested):
I considered making my comments non-blocking but documentation page really needs some love :) thanks!
